sklearn/ensemble/_forest.py

```diff
@@ -624,17 +664,31 @@ def _set_oob_score_and_attributes(self, X, y):
             The data matrix.
         y : ndarray of shape (n_samples, n_outputs)
             The target matrix.
+        sample_weight : ndarray of shape (n_samples,)
+            Sample weights.
         """
-        self.oob_decision_function_ = super()._compute_oob_predictions(X, y)
+        self.oob_decision_function_, oob_importances_ = (
+            super()
+            ._compute_oob_predictions_and_importances(X, y, sample_weight)
+        )
+
         if self.oob_decision_function_.shape[-1] == 1:
             # drop the n_outputs axis if there is a single output
             self.oob_decision_function_ = self.oob_decision_function_.squeeze(
                 axis=-1
             )
         self.oob_score_ = accuracy_score(
-            y, np.argmax(self.oob_decision_function_, axis=1)
+            y, np.argmax(self.oob_decision_function_, axis=1),
+            sample_weight=sample_weight
         )
 
+        if self.feature_importances_type == "permutation":
```

Comment:
I am proposing the following patch avoiding the duplication for setting the permutation importance in the specialized method:

<details>

```diff
diff --git a/sklearn/ensemble/_forest.py b/sklearn/ensemble/_forest.py
index 7bbfdcf320..6997382859 100644
--- a/sklearn/ensemble/_forest.py
+++ b/sklearn/ensemble/_forest.py
@@ -432,6 +432,11 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
     def _set_oob_score_and_attributes(self, X, y, sample_weight):
         """Compute and set the OOB score and attributes.
 
+        This abstract method compute the predictions and the importances
+        (if required) and will set the permutation importance attributes.
+        The predictions are returned to the child methods to set the
+        specialized class attributes.
+
         Parameters
         ----------
         X : array-like of shape (n_samples, n_features)
@@ -441,6 +446,19 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
         sample_weight : ndarray of shape (n_samples,)
             Sample weights.
         """
+        oob_predictions, oob_importances = \
+            self._compute_oob_predictions_and_importances(
+                X, y, sample_weight
+            )
+
+        if oob_importances is not None:
+            self._oob_permutation_importance = Bunch(
+                importances_mean=np.mean(oob_importances, axis=1),
+                importances_std=np.std(oob_importances, axis=1),
+                importances=oob_importances
+            )
+
+        return oob_predictions
 
     def _compute_oob_predictions_and_importances(self, X, y, sample_weight):
         """Compute and set the OOB score.
@@ -667,9 +685,8 @@ class ForestClassifier(ClassifierMixin, BaseForest, metaclass=ABCMeta):
         sample_weight : ndarray of shape (n_samples,)
             Sample weights.
         """
-        self.oob_decision_function_, oob_importances_ = (
-            super()
-            ._compute_oob_predictions_and_importances(X, y, sample_weight)
+        self.oob_decision_function_ = super()._set_oob_score_and_attributes(
+            X, y, sample_weight
         )
 
         if self.oob_decision_function_.shape[-1] == 1:
@@ -677,16 +694,11 @@ class ForestClassifier(ClassifierMixin, BaseForest, metaclass=ABCMeta):
             self.oob_decision_function_ = self.oob_decision_function_.squeeze(
                 axis=-1
             )
-        self.oob_score_ = accuracy_score(
-            y, np.argmax(self.oob_decision_function_, axis=1),
-            sample_weight=sample_weight
-        )
 
-        if self.feature_importances_type == "permutation":
-            self._oob_permutation_importance = Bunch(
-                importances_mean=np.mean(oob_importances_, axis=1),
-                importances_std=np.std(oob_importances_, axis=1),
-                importances=oob_importances_
+        if self.oob_score:
+            self.oob_score_ = accuracy_score(
+                y, np.argmax(self.oob_decision_function_, axis=1),
+                sample_weight=sample_weight
             )
 
     def _validate_y_class_weight(self, y):
@@ -974,23 +986,17 @@ class ForestRegressor(RegressorMixin, BaseForest, metaclass=ABCMeta):
         sample_weight : ndarray of shape (n_samples,)
             Sample weights.
         """
-        self.oob_prediction_, oob_importances_ = (
-            super()
-            ._compute_oob_predictions_and_importances(X, y, sample_weight)
+        self.oob_prediction_ = super()._set_oob_score_and_attributes(
+            X, y, sample_weight
         )
         self.oob_prediction_ = self.oob_prediction_.squeeze(axis=1)
         if self.oob_prediction_.shape[-1] == 1:
             # drop the n_outputs axis if there is a single output
             self.oob_prediction_ = self.oob_prediction_.squeeze(axis=-1)
-        self.oob_score_ = r2_score(
-            y, self.oob_prediction_, sample_weight=sample_weight
-        )
 
-        if self.feature_importances_type == "permutation":
-            self._oob_permutation_importance = Bunch(
-                importances_mean=np.mean(oob_importances_, axis=1),
-                importances_std=np.std(oob_importances_, axis=1),
-                importances=oob_importances_
+        if self.oob_score:
+            self.oob_score_ = r2_score(
+                y, self.oob_prediction_, sample_weight=sample_weight
             )
 
     def _compute_partial_dependence_recursion(self, grid, target_features):
```

Reply:
In this way, we only specialized the predictions/decision_function and score for each class.